core/src/identifiers.rs

```diff
@@ -0,0 +1,190 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Base identifiers for messaging primitives.
+
+use crate::message::ExitCode;
+use alloc::vec::Vec;
+use blake2_rfc::blake2b;
+
+/// Creates a unique identifier by passing given argument to blake2b hash-function.
+fn hash(argument: &[u8]) -> [u8; 32] {
+    let mut arr = [0; 32];
+
+    let blake2b_hash = blake2b::blake2b(32, &[], argument);
+    arr[..].copy_from_slice(blake2b_hash.as_bytes());
+
+    arr
+}
+
+/// Declares data type for storing any kind of id for gear-core,
+/// which stores 32 bytes under the hood.
+macro_rules! declare_id {
+    ($name:ident: $doc: literal) => {
+        #[doc=$doc]
+        #[derive(
+            Clone,
+            Copy,
+            Debug,
+            Default,
+            Eq,
+            Hash,
+            Ord,
+            PartialEq,
+            PartialOrd,
+            codec::Decode,
+            codec::Encode,
+            derive_more::From,
+            scale_info::TypeInfo,
+        )]
+        pub struct $name([u8; 32]);
+
+        impl From<$name> for [u8; 32] {
+            fn from(val: $name) -> [u8; 32] {
+                val.0
+            }
+        }
+
+        impl From<u64> for $name {
```

Comment:
I totally agree with you, but this will incredible inflate the PR.

We use this from-into in many tests and many crates. While we build for tests crate A, which depend on crate B, crate B doesn't build with test feature.

I'm going to remove all this redundant from-into for ids in future PRs. I hope we can lead the code to condition when we build ids only once and only with new generation-rules (all other methods to build these structs will be deprecated)